pip/pip-420.md

```diff
@@ -0,0 +1,276 @@
+  
+    for (int i = 0; i < 10; i++) {
+      producer.send(new User("name-" + i, 10 + i));
+    }
+  
+    for (int i = 0; i < 10; i++) {
+      Message<User> message = consumer.receive();
+      consumer.acknowledge(message);
+      assertEquals(message.getValue().getName(), "name-" + i);
+    }
+}
+```
+
+## Public-facing Changes
+
+Introduce a new SchemaType `EXTERNAL` to represent the schema types that work with external schema registry.
+```java
+public enum SchemaType {
+
+    /**
+     * External Schema Type.
+     * <p>
+     * This is used to indicate that the schema is managed externally, such as in a schema registry.
+     * External schema type is not compatible with any other schema type.
+     * </p>
+     */
+    EXTERNAL(21)
+
+}
+```
+
+```protobuf
+// File `SchemaRegistryFormat.proto`
+message SchemaInfo {
+    enum SchemaType {
+        EXTERNAL = 22;
+    }
+}
+```
+
+```protobuf
+// File `PulsarApi.proto`
+message Schema {
+    enum Type {
+        External = 22;
+    }
+}
+```
+
+Add a new method `getConfigs` for `SchemaInfoProvider` interface to provide necessary params for external schemas connect to the schema registry service.
+```java
+public interface SchemaInfoProvider {
+
+    /**
+      * Returns the configs of the schema registry service, such as URL, authentication params.
+      */
+    default Map<String, String> getConfigs() {
+        return Collections.emptyMap();
+    }
+
+}
+```
+
+The client build supports setting the `schemaProperties`.
+```java
+public interface ClientBuilder extends Serializable, Cloneable {
+
+  /**
+   * Set the properties used for schema.
+   * <p>
+   *     These properties will be used to configure the schema registry client.
+   * </p>
+   * @param properties schema registry properties
+   */
+  ClientBuilder schemaProperties(Map<String, String> properties);
+
+}
+```
+
+The `ClientConfigurationData` supports transfer `schemaProperties`.
+```java
+public class ClientConfigurationData implements Serializable, Cloneable {
+
+    private Map<String, String> schemaProperties;
+
+}
+```
+
+The customized external schemas can get the `SchemaInfoProvider` and retrieve the configs from it, extends the interface `AutoCloseable` to support close external schema resources.
+```java
+public interface Schema extends Cloneable, AutoCloseable {
+
+    @Override
+    default void close() {
+        // no-op
+    }
+
+    /**
+     * When setting schema info provider for schema, the schema can retrieve the configs and initialize itself.
+     */
+    default void setSchemaInfoProvider(SchemaInfoProvider schemaInfoProvider) {
+    }
+
+}
+```
+
+# Pulsar Function
+
+For support using third-party schema registry service in Pulsar Function,
+- Support setting the `schemaProperties` while initializing the Pulsar client
+- Support the `SchemaType.EXTERNAL` schema type in Pulsar Function
+
+# Security Considerations
+
+Users can provide schema registry security configuration in the `schemaProperties`.
```

Comment:
Since the client directly connects to the external schema store, does this proposal involve the definition of the authentication and authorization interface?

Reply:
No, users just need to provide security configurations to the configuration `schemaProperties`, it's used to initialize the external schema registry client.